src/main/java/org/folio/dataimport/handlers/actions/CreateInvoiceEventHandler.java

```diff
@@ -108,41 +113,54 @@ public CompletableFuture<DataImportEventPayload> handle(DataImportEventPayload d
-            InvoiceLineCollection invoiceLineCollection = new InvoiceLineCollection().withInvoiceLines(invoiceLines).withTotalRecords(invoiceLines.size());
-            dataImportEventPayload.getContext().put(INVOICE_LINES_KEY, Json.encode(invoiceLineCollection));
-            Map<Integer, String> invoiceLinesErrors = prepareInvoiceLinesErrors(result.result());
-            if (!invoiceLinesErrors.isEmpty()) {
-              dataImportEventPayload.getContext().put(INVOICE_LINES_ERRORS_KEY, Json.encode(invoiceLinesErrors));
-              future.completeExceptionally(new EventProcessingException("Error during invoice lines creation"));
-              return;
-            }
-            future.complete(dataImportEventPayload);
-          } else {
-            preparePayloadWithMappedInvoiceLines(dataImportEventPayload);
-            logger.error("Error during invoice creation", result.cause());
-            future.completeExceptionally(result.cause());
-          }
         });
+      var recordId = payloadContext.get(RECORD_ID);
+      Future<RecordToEntity> recordToInvoiceFuture = idStorageService.store(recordId, UUID.randomUUID().toString(),
+        dataImportEventPayload.getTenant());
+
+      recordToInvoiceFuture.onSuccess(res -> {
+        String invoiceId = res.getEntityId();
+        invoicesFuture
+          .compose(v -> saveInvoice(dataImportEventPayload, okapiHeaders, invoiceId))
+          .map(savedInvoice -> prepareInvoiceLinesToSave(savedInvoice.getId(), dataImportEventPayload, poLinesFuture.result()))
+          .compose(preparedInvoiceLines -> saveInvoiceLines(preparedInvoiceLines, okapiHeaders))
+          .onComplete(result -> {
+            makeLightweightReturnPayload(dataImportEventPayload);
+
+            if (result.succeeded()) {
+              List<InvoiceLine> invoiceLines = result.result().stream().map(Pair::getLeft).collect(Collectors.toList());
+              InvoiceLineCollection invoiceLineCollection = new InvoiceLineCollection().withInvoiceLines(invoiceLines).withTotalRecords(invoiceLines.size());
+              dataImportEventPayload.getContext().put(INVOICE_LINES_KEY, Json.encode(invoiceLineCollection));
+              Map<Integer, String> invoiceLinesErrors = prepareInvoiceLinesErrors(result.result());
+              if (!invoiceLinesErrors.isEmpty()) {
+                dataImportEventPayload.getContext().put(INVOICE_LINES_ERRORS_KEY, Json.encode(invoiceLinesErrors));
+                future.completeExceptionally(new EventProcessingException("Error during invoice lines creation"));
+                return;
+              }
+              future.complete(dataImportEventPayload);
+            } else {
+              preparePayloadWithMappedInvoiceLines(dataImportEventPayload);
+              logger.error("Error during invoice creation", result.cause());
+              future.completeExceptionally(result.cause());
+            }
+          });
+      }).onFailure(failure -> {
```

Comment:
we should fail the feature with DuplicateEventException in case a duplicate event has been received.
In the saveInvoice() method it's needed to check if the error occurs about invoice id duplication then we need to return future failed by DuplicateEventException.